group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -583,11 +674,330 @@ private void unload() {
+                        currentTimeMs,
+                        serializer.serializeKey(record),
+                        serializer.serializeValue(record)
+                    ));
+                }
+
+                // Compute the estimated size of the records.
+                int estimatedSize = AbstractRecords.estimateSizeInBytes(
+                    currentBatch.builder.magic(),
+                    compression.type(),
+                    recordsToAppend
+                );
+
+                // Check if the current batch has enough space. We check is before
+                // replaying the records in order to avoid having to revert back
+                // changes if the records do not fit within a batch.
+                if (!currentBatch.builder.hasRoomFor(estimatedSize)) {
+                    if (currentBatch.builder.numRecords() == 0) {
+                        // If the number of records in the current batch is zero, it means that
+                        // the records are larger than the max batch size.
+                        throw new RecordTooLargeException("Message batch size is " + estimatedSize +
+                            " bytes in append to partition " + tp + " which exceeds the maximum " +
+                            "configured size of " + currentBatch.maxBatchSize + ".");
+                    } else {
+                        // Otherwise, we write the current batch, allocate a new one and re-verify
+                        // whether the records fit in it.
+                        writeCurrentBatch();
+                        maybeAllocateNewBatch(
+                            producerId,
+                            producerEpoch,
+                            verificationGuard,
+                            currentTimeMs
+                        );
+                        if (!currentBatch.builder.hasRoomFor(estimatedSize)) {
+                            throw new RecordTooLargeException("Message batch size is " + estimatedSize +
+                                " bytes in append to partition " + tp + " which exceeds the maximum " +
+                                "configured size of " + currentBatch.maxBatchSize + ".");
+                        }
+                    }
+                }
+
+                // Add the event to the list of pending events associated with the batch.
+                currentBatch.events.add(event);
+
+                try {
+                    // Apply record to the state machine.
+                    if (replay) {
+                        for (int i = 0; i < records.size(); i++) {
+                            // We compute the offset of the record based on the last written offset. The
+                            // coordinator is the single writer to the underlying partition so we can
+                            // deduce it like this.
+                            coordinator.replay(
+                                currentBatch.nextOffset + i,
```

Comment:
for my understanding -- this used to be prevLastWrittenOffset.
So the currentBatch.nextOffset value is equivalent? Seems like these values would be off by one, but maybe that's just a naming thing.

Reply:
the naming seems a bit off. nextOffset is initialized as baseOffset and we increment it below which means it's pointing to LEO - 1 right?

Reply:
The last written offset is essentially the log end offset after a batch is written. It is the offset until which we must wait in order to consider the state related to the batch committed. When a new batch is created, its start offset is set to the last written offset (the LEO) and its next offset is set to the same value.

When records are replayed, we can no longer use the last written offset to compute the offset of the record because we must take into account the records pending in the pending batch.

I agree that "last written offset" is probably a poor name. I wonder if we should use "log end offset" in place. I could also rename `nextOffset` to `endOffset`. I am open to suggestions here.

Reply:
Yeah, last written / next were the parts confusing to me.
I think I also saw "current batch" as the one we were replaying/going to write next, but it's actually the last replicated batch in the log? and pending batch is the one that is going to be written.

Reply:
@jolshan You're correct. The current batch (or pending batch) is the one that we are writing to and that is NOT written yet.

Reply:
Ok -- my last comment was slightly wrong then. I understand what you mean now, and the last thing I was confused about was that we aren't writing anything to the log, yet but adjusting the values since we haven't appended these i records to the batch yet.